library/alloc/src/slice.rs

```diff
@@ -1036,149 +1040,150 @@ where
 
     impl<T> Drop for MergeHole<T> {
         fn drop(&mut self) {
-            // `T` is not a zero-sized type, so it's okay to divide by its size.
-            let len = (self.end as usize - self.start as usize) / mem::size_of::<T>();
             unsafe {
+                // use slice method for length calculation
+                let len = self.end.offset_from(self.start) as usize;
```

Comment:
Does this line need to move inside the `unsafe` block?

Reply:
Yes, otherwise:
![image](https://user-images.githubusercontent.com/6633152/140320340-4f64933a-cbe9-416e-aba6-ada79e1ffaeb.png)

Reply:
Yes, but was it strictly needed to change to `offset_from`?

Reply:
Not strictly needed, but I think it is better to use a slice method instead of "rolling your own" as it was before:
```
            // `T` is not a zero-sized type, so it's okay to divide by its size.
            let len = (self.end as usize - self.start as usize) / mem::size_of::<T>();
```

Reply:
I'm not so sure about that, actually. This is the Rust standard library. The source code doesn't have to be pretty if it can't be, but safety very important. Every unsafe statement adds to the risk of a defect creeping in.

Reply:
I prefer to use the standard slice method.  The original code relies on knowledge of the underlying layout/representation.